### PR TITLE
Continue committing received transactions on leader standdown

### PR DESCRIPTION
**Pull request by tylerkaraszewski**
## Fixes:
$ https://github.com/Expensify/Expensify/issues/143561

When leader stands down, followers go SEARCHING. This means they stop committing replicated transactions until a new leader stands up.

However, with multi-rep and asynchronous replication, it's entirely possible that the following happens:

1. Leader broadcasts transaction.
2. Follower begins transaction.
3. Leader stands down.
4. Follower abandons transactions in progress.

It actually happens often enough that the final transaction sent be leader isn't committed *anywhere* else because every other node gets the `stand down` message and abandons all in-progress transactions before that last transaction is committed. The slower that transaction happens to be, the less likely it is to get committed. If no other nodes commit it, then leader has forked as it shut down one transaction ahead of the rest of the cluster.

This change adds a parameter to `cancel` on a sequential notifier to only cancel transactions *after* whatever leader has told us its most recent commit is. This means that before falling out of FOLLOWING, we'll catch up to wherever we think leader was when it shut down. Note that this only applies to messages we've already deserialized from leader, so if we're *really* far behind, we'll just finish whatever we've already started, but we do not start any new replication transactions.

## Tests
This has run through `clustertest` 20x in a row with no failures.

### CODE REVIEW

**Comment by righdforsa** [comment]
I'm not sure if it's a cleaner solution, as I'm doing my best to understand what is happening in this PR, but wouldn't we want part of the leader standing down be waiting for commit confirmations to come in, rather than having followers rely on their idea of where leader was?

**Comment by tylerkaraszewski** [comment]
We need something that works reasonably well when leader crashes as well. We could refuse to stand down until at least one follower said it was on the same commit as leader, though. But I don't think that'll get us any further than this PR does.

**Comment by tylerkaraszewski** [comment]
To expand, let's say leader sends a BEGIN and a COMMIT.

These received by a follower running the code in this PR. Leader shuts down before the COMMIT runs, but with this change, the COMMIT does still run.

Let's say we modified the leading code that the leader sends a BEGIN and then a COMMIT, and followers do the BEGIN, do the COMMIT, and then on their next message to leader, have a high-enough commit count that leader can shut down.

The end result is identical.

Ok, let's consider then some error case:

Leader sends a BEGIN and a COMMIT. This is for a non-quorum transaction, as all of this stuff is already handled for a quorum transaction.

Followers receive the BEGIN and the COMMIT, or they probably do, how does leader know for sure? Anyway, they either get them or they don't, but they never send a message back to leader indicating that they committed that transaction. Maybe that transaction caused them to all crash? Maybe they did some other thing that's weird?

Ok then, leader doesn't get a response, because none of its followers are committing its transaction. What's it supposed to do now? Everything is screwed at this point already, and the state is unrecoverable anyway.

**Comment by righdforsa** [comment]
Hmm, I'm still confused, bear with me. Crashes are not explicitly what you're trying to fix, here, right? If you want to fix something that works for crashes too, then why wouldn't we just process all outstanding transaction requests any time we lose the leader for any reason? I would guess that's as "nearly safe" as my unsubstantiated fears about this.

Afaict, follower behavior doesn't need to change. If a leader crashes, it will either be before or after it sent the COMMIT message for its last transaction or it won't.
1. if a follower processes the COMMIT before it realizes leader is gone, it will commit, nbd, normal operation
2. if a follower doesn't process the COMMIT before it realizes leader is gone, it will abandon the commit and rollback
in either case the follower will synchronize correctly when it goes searching. 

The problem is entirely on leader expectations when shutting down. If it crashes, it's a dice roll. Either it:
1. happened to commit right before the crash, in which case, we know that's a possible leader fork condition (unless we implement a "process all queued transaction commands" as the standard operating posture, mentioned above)
2. it didn't finish committing in which case sqlite should hopefully keep that state, and it should rollback when the process starts back up.

However, we have control over a shutdown, and we have an existing mechanism to enforce consistency so, wouldn't it just be super simple once the leader receives a shutdown to set the quorum threshold on the next commit, and then the standard mechanism comes into play? Or should the proposal be to move to processing all outstanding transaction commands (I suppose until we get a hash mismatch or other commit failure) before dropping to searching?

File: sqlitecluster/SQLiteSequentialNotifier.cpp
Comment (righdforsa) [inline sqlitecluster/SQLiteSequentialNotifier.cpp:67]: NAB: existing typo "fond"

**Comment by rafecolton** [comment]
> if a follower doesn't process the COMMIT before it realizes leader is gone, it will abandon the commit and rollback
in either case the follower will synchronize correctly when it goes searching.

If the follower doesn't process the commit but the leader does, and then the leader crashes, and meanwhile, another leader takes over and commits, when the previous leader comes back up, it will have forked.

**Comment by tylerkaraszewski** [comment]
You are misunderstanding the problem.

> if a follower doesn't process the COMMIT before it realizes leader is gone, it will abandon the commit and rollback
in either case the follower will synchronize correctly when it goes searching.

It will absolutely not synchronize correctly, because the follower discarded the commit it needed to sync. 

Imagine a 3-node cluster with all nodes currently at commit #100. Node 0 is leading. Node 1 and 2 are following.

The following occurs in this order:

1. Node 0 commits transaction 101.
2. Nodes 1 and 2 receive transaction 101 and begin to run it.
3. Node 0 stops leading (gracefully or not, it doesn't matter)
4. Nodes 1 and 2 roll back transaction 101, as they have lost their leader.
5. Node 1 begins leading. It commits a new and different transaction 101. 
6. Node 2 replicates the new and different transaction 101 from node 1.
7. Node 0 is restarted and comes back up and attempts to rejoin the cluster but can't, because it's commit 101 doesn't match the rest of the cluster.

The problem is step 4 above. If leader has committed a transaction, and a follower knows about that transaction, it should commit it as well. The leader can shutdown gracefully, or it can crash, but if the followers willingly discard transactions it sent them before shutting down, then it's likely to fork. Followers need to commit transactions once received from leader.

**Review by rafecolton — APPROVED** [review]
Reasoning through this, it does seem like it will help reduce the likelihood of the leader-fork-at-shutdown scenario we've been seeing. It would be great to see this demonstrated with a test. Discussed with @tylerkaraszewski 1:1 for how this test would work and why it should't be a permanent addition to the test suite:

> So, the only way I can think of to reproduce it requires hacking the replication code to take a long time to run queries. If it takes 10 seconds to commit a replicated query, we can pretty reliably do a commit on leader, and then disconnect before followers finish that commit.
> 
> I can probably do that with a slow-on-purpose test command, but even then it's all timing based, and slow on purpose. I could potentially do it as a one-off, and validate that it failed reliably on the old code, but passed in the new code, but it's going to be one long slow test for an incredibly edge-case scenario if we add it to the regular test suite.

Approving and will wait for the demo before merging.

**Comment by agzakula** [comment]
I think I totally understand the failure case. If the only problem we're
trying to solve is the leader shutting down after committing a request that
hasn't been replicated, the solution is much more simple: all we have to do
is set the full quorum for the last commit before standing down, and wait
for quorum to be achieved.

Alternatively, if we want a solution that allows followers to commit
transactions sent by a leader before it went away, wouldn't it be simpler
and not much less safe (i.e. equally unsafe) to just always process any
outstanding transaction commands before going looking for another leader?
Keeping track of the leader's commit count is always going to be
out-of-date, and it doesn't really matter.

**Comment by tylerkaraszewski** [comment]
> all we have to do is set the full quorum for the last commit before standing down, and wait for quorum to be achieved.

What commit is that? The next one to come in after we begin shutting down? This doesn't work on an idle server. I know, our production servers are never idle, but it would break any open source users, and also all of our tests and VMs.

If the last commit is done with QUORUM, then nothing is outstanding, yes, but there's no such thing as a "last commit".

> just always process any outstanding transaction commands before going looking for another leader?

We can only process transactions that we know will be committed. I.e., commands with a commit count <= leader's current commit count. Otherwise leader can send a `BEGIN` and then crash (or we get disconnected, or whatever) and we'll sit there forever trying to finish a transaction that will never have a COMMIT come. The solution is to commit all the transactions that we know can be committed - i.e., up to leader's current commit count.

The problem here is "we don't commit transactions that we could" and the solution is "commit all the transactions we know we can".

**Comment by tylerkaraszewski** [comment]
Also, agzakula is back.

**Comment by tylerkaraszewski** [comment]
I'm testing this change with the following test code:
```
#include "../BedrockClusterTester.h"

struct ShutdownReplicationTest : tpunit::TestFixture {
    ShutdownReplicationTest()
        : tpunit::TestFixture("ShutdownReplication", TEST(ShutdownReplicationTest::test)) { }

    uint64_t getNodeCommitCount(BedrockClusterTester& tester, size_t nodeIndex) {
        SData status("Status");
        auto json = SParseJSONObject(tester.getTester(nodeIndex).executeWaitMultipleData({status}, 1, true)[0].content);
        return stoull(json["commitCount"]);
    }

    void test()
    {
        // Create a cluster, let it stand up.
        BedrockClusterTester tester;

        // Make sure all commit counts are 1, which will be the commit for the `upgradeDatabase` call.
        ASSERT_EQUAL(getNodeCommitCount(tester, 0), 1);
        ASSERT_EQUAL(getNodeCommitCount(tester, 1), 1);
        ASSERT_EQUAL(getNodeCommitCount(tester, 2), 1);

        // Now send a slow-writing command to leader and then immediately shut down.
        // This should increment the entire cluster to a commitCount of 2.
        SData slowWrite("slowprocessquery");
        slowWrite["size"] = "1000";
        slowWrite["count"] = "1000";
        tester.getTester(0).executeWaitMultipleData({slowWrite});
        tester.getTester(0).stopServer();

        // Wait until the first follower is leading, and see if the cluster received the commit.
        ASSERT_TRUE(tester.getTester(1).waitForState("LEADING"));
        ASSERT_EQUAL(getNodeCommitCount(tester, 1), 2);
        ASSERT_EQUAL(getNodeCommitCount(tester, 2), 2);
    }
} __ShutdownReplicationTest;
```

I've verified it fails on the master branch:
```
 clustertest:$  ./clustertest -only ShutdownReplication
[--------------]
[ RUN          ] ShutdownReplicationTest::test
[              ]    assertion #1 at test/clustertest/tests/ShutdownReplicationTest.cpp:33
[       FAILED ] ShutdownReplicationTest::test
[--------------]

[==============]
[ TEST RESULTS ] Passed: 0, Failed: 1
[==============]
 clustertest:$  git status
On branch master
```

And that is passes with this branch:
```
 clustertest:$  ./clustertest -only ShutdownReplication
[--------------]
[ RUN          ] ShutdownReplicationTest::test
[       PASSED ] ShutdownReplicationTest::test
[--------------]

[==============]
[ TEST RESULTS ] Passed: 1, Failed: 0
[==============]
 clustertest:$  git status
On branch tyler-continue-committing-received-transactions-on-leader-standdown
```

**Review by coleaeason — APPROVED** [review]

**Comment by coleaeason** [comment]
this lgtm, unless @righdforsa is going to add a blocking review, let's merge this bad boy

**Pull request merged**
